File: 그래프/boj_11725.py
import sys
from collections import defaultdict
input = sys.stdin.readline


def find_parent(n, edges):
    graph = defaultdict(list)
    parents = [0] * (n+1)

    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)

    visited = [False] * (n+1)
    stack = [1]

    while stack:
        node = stack.pop()

        for adj in graph[node]:
            if not visited[adj]:
                parents[adj] = node
                stack.append(adj)
                visited[adj] = True

    return parents[2:]


def main():
    n = int(input())
    edges = [list(map(int, input().split())) for _ in range(n-1)]

    result = find_parent(n, edges)
    for parent in result:
        print(parent)


if __name__ == "__main__":
    main()

File: 그래프/boj_11725.java
package 그래프;
import java.io.*;
import java.util.*;

public class boj_11725 {
    static int N;
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N + 1; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < N - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());            
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        int[] parent = findParents();
        for (int i = 2; i < N + 1; i++) {
            bw.write(parent[i] + "\n");
        }
        bw.close();
    }

    static int[] findParents() {
        boolean[] visited = new boolean[N + 1];
        int[] parent = new int[N + 1];

        ArrayDeque<Integer> stack = new ArrayDeque<>();
        stack.push(1);
        visited[1] = true;

        while (!stack.isEmpty()) {
            int node = stack.pop();

            for (int next : graph.get(node)) {
                if (!visited[next]) {
                    parent[next] = node;
                    stack.push(next);
                    visited[next] = true;
                }
            }
        }
        return parent;
    }    
}
